src/passes/Poppify.cpp

```diff
@@ -0,0 +1,504 @@
+
+  // Maps tuple locals to the new locals that will hold their elements
+  std::unordered_map<Index, std::vector<Index>> tupleVars;
+
+  // Records the scratch local to be used for tuple.extracts of each type
+  std::unordered_map<Type, Index> scratchLocals;
+
+  // Maps labels to the types they receive. Used for typing breaks.
+  // TODO: is this necessary?
+  std::unordered_map<Name, Type> labelTypes;
+
+  Poppifier(Function* func, Module* module);
+
+  Index getScratchLocal(Type type);
+
+  // Replace `expr`'s children with Pops of the correct type.
+  void poppify(Expression* expr);
+
+  // Replace `expr` with a block containing the instructions for the current
+  // scope and clear the current scope.
+  void patchInstrs(Expression*& expr);
+
+  // BinaryIRWriter methods
+  void emit(Expression* curr);
+  void emitHeader() {}
+  void emitIfElse(If* curr);
+  void emitCatch(Try* curr, Index i);
+  void emitCatchAll(Try* curr);
+  void emitScopeEnd(Expression* curr);
+  void emitFunctionEnd();
+  void emitUnreachable();
+  void emitDebugLocation(Expression* curr) {}
+
+  // Tuple lowering methods
+  void emitTupleExtract(TupleExtract* curr);
+  void emitDrop(Drop* curr);
+  void emitLocalGet(LocalGet* curr);
+  void emitLocalSet(LocalSet* curr);
+  void emitGlobalGet(GlobalGet* curr);
+  void emitGlobalSet(GlobalSet* curr);
+};
+
+Poppifier::Poppifier(Function* func, Module* module)
+  : BinaryenIRWriter<Poppifier>(func), module(module), builder(*module) {
+  // Start with a scope to emit top-level instructions into
+  scopeStack.emplace_back(Scope::Func);
+
+  // Map each tuple local to a set of expanded locals
+  for (Index i = func->getNumParams(), end = func->getNumLocals(); i < end;
+       ++i) {
+    Type localType = func->getLocalType(i);
+    if (localType.isTuple()) {
+      auto& vars = tupleVars[i];
+      for (auto type : localType) {
+        vars.push_back(builder.addVar(func, type));
+      }
+    }
+  }
+}
+
+Index Poppifier::getScratchLocal(Type type) {
+  // If there is no scratch local for `type`, allocate a new one
+  auto insert = scratchLocals.insert({type, Index(-1)});
+  if (insert.second) {
+    insert.first->second = builder.addVar(func, type);
+  }
+  return insert.first->second;
+}
+
+void Poppifier::patchInstrs(Expression*& expr) {
+  auto& instrs = scopeStack.back().instrs;
+  if (auto* block = expr->dynCast<Block>()) {
+    // Reuse blocks, but do not patch a block into itself, which would otherwise
+    // happen when emitting if/else or try/catch arms and function bodies.
+    if (instrs.size() == 0 || instrs[0] != block) {
+      block->list.set(instrs);
+    }
+  } else {
+    // Otherwise create a new block, even if we have just a single
+    // expression. We want blocks in every new scope rather than other
+    // instructions because Poppy IR optimizations only look at the children of
+    // blocks.
+    expr = builder.makeBlock(instrs, expr->type);
+  }
+  instrs.clear();
```

Comment:
Yes, because we don't always pop the `Scope`. For example, in `emitIfElse` we keep the scope but change its `kind`. If we didn't do an `instrs.clear()` here, the instructions from the first arm would incorrectly also appear in the second arm.

Reply:
I see. I personally think treating `Scope` like unit in that in `emitIfElse` we pop an old scope and push a new scope, is a little more intuitive, but I'm not very opinionated on that.

Reply:
Ack. I like that `patchInstrs` clears the instructions because each instruction should logically be "consumed" by `patchInstrs` exactly once, so having `patchInstrs` actually consume them by clearing them from the scope stack ensures that this logic cannot be messed up. I think I can take this farther and actually have `patchInstrs` pop the entire scope so it matches your intuition as well.